### PR TITLE
Godeps, vendor: switch to Go 1.6 vendored dependencies

### PR DESCRIPTION
**Pull request by karalabe**
This PR finally moves all of our dependencies to the `vendor` folder. This should enable our repository to be go gettable, and most importantly, reusable as a library in other Go projects without needing to jump though insane dependency management hoops. Yay! :tada: 

*Note: with the introduction of the vendor folder we can no longer do `go install ./...`, as that would install everything from `vendor` too, most of which are junk which we don't vendor nor use. This is a [known wontfix](https://github.com/golang/go/issues/11659) issue. I've updated the `Makefile` accordingly to skip the vendor folder when building and only use it for the dependencies, not as main codebase.*

### CODE REVIEW

**Comment by robotally** [comment]
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Tue May 3 16:27:15 UTC 2016_

**Comment by karalabe** [comment]
@fjl Please review this PR, especially my changes to the makefile and build env files.

The CI servers need to be updated to Go 1.6, otherwise they won't pass. @caktux Could you please update all CI servers to Go 1.6.2?

File: node/node.go [hunk 1]
```diff
@@ -152,7 +152,7 @@ func (n *Node) Start() error {
 	}
 	// Otherwise copy and specialize the P2P configuration
 	running := new(p2p.Server)
-	*running = *n.serverConfig
+	reflect.ValueOf(running).Elem().Set(reflect.ValueOf(n.serverConfig).Elem()) // Fancy *running = *n.serverConfig to prevent mutex copy
```
Comment (fjl) [inline node/node.go]: ??

Comment (karalabe) [inline node/node.go]: https://travis-ci.org/ethereum/go-ethereum/builds/127570769#L301

Comment (fjl) [inline node/node.go]: I understand the intention to fix the `go vet` issue but this is not a good fix for it.
We need to find another way. I can remove the mutex as a quick workaround.

**Comment by fjl** [comment]
> as that would install everything from vendor too, most of which are junk which we don't vendor nor use.

That's weird. Why do we vendor 'junk' that is not used? If godep leaves unused dependencies in the
directory we should switch to a different tool instead.

**Comment by karalabe** [comment]
Well, for starters, the windows pipe package isn't build constrained to windows only, so every other platform will fail to build `./...`. Only the windows RPC files reference it, so it is only touched on Windows by our code, but `./...` doesn't know that.

**Comment by fjl** [comment]
That sounds like an issue we can fix though.
I am arguing to remove the (go list) workaround because
it highlights such issues.

**Comment by karalabe** [comment]
I understand your point, but do tell me how you think we can solve it? I don't think doing manual patches to upstream dependencies is such a good idea.

**Comment by karalabe** [comment]
Another example that breaks is @Gustav-Simonsson 's opencl dependency, since even though ethash has a build tag/constrain to only use it if requested, when doing `./...` we inherently build it, which fails if you don't have the OpenCL develop libs available.

**Comment by caktux** [comment]
This affects more than just the golang version... Build server steps, Homebrew formula and deb packaging scripts will all need to be changed to support this, and it also means we'll have to yet again split the build procedures for master and develop like the good ol' PoC days. So this will need a little more planning besides just getting the builds to pass again.

**Comment by fjl** [comment]
The way to build reliably is to run `make all`. I have updated the Homebrew formula recently to do just that. Can we update the buildbot steps, deb scripts to run make all too?

**Comment by caktux** [comment]
What about Windows? I think I've got the others ready at least for the buildslaves with 1.6 and the build steps with `make all`.

**Comment by fjl** [comment]
It's true that we can't build with `make` on Windows.

I'd say the easiest way would be to add a .bat script to go-ethereum that does the build on both master and develop. Compared to the make build it would still require setting up `GOPATH` and cloning to the right place but that's fine. We only need it so buildbot can run simple steps that we can tweak through the script. I'll look into adding it.

**Comment by caktux** [comment]
@karalabe `make all` is having issues with the ARM builds, something about GOBIN being set in [those logs](https://builds.ethereum.org/builders/ARM%20Go%20pull%20requests/builds/2497/steps/make-all/logs/stdio).. since you're the cross-compilation expert, could you give a hand with that builder? Linux and OSX seem to be doing great.

So besides ARM and Windows, all that should be left is the deb packaging (the `rules` file more specifically).

@tgerring Could you update golang to 1.6 on the Windows buildslave?

**Comment by karalabe** [comment]
@caktux I've souped up our makefile in this PR https://github.com/ethereum/go-ethereum/pull/2520 to handle the cross build issue of Go, but I hit a gcc issue in your arm images: https://builds.ethereum.org/builders/ARM%20Go%20pull%20requests/builds/2501/steps/go-test/logs/stdio . You probably need to sort this out on your side.

**Comment by caktux** [comment]
Fixed, wasn't passing the env, but then we obviously can't test cross-compiled binaries either so just removed the test step as it was before. Once #2520 is merged it might be good to rebase this one.

**Comment by fjl** [comment]
@caktux Windows scripts are present on both master and develop now.

**Comment by VoR0220** [comment]
👍

**Comment by fjl** [comment]
We'll redo this change in the new build/ci.go script when Go 1.7 is out and available for most distros.

**Pull request closed without merging**